### PR TITLE
feat: referrals

### PR DESCRIPTION
**Pull request by enthusiastmartin**
Adds referrals pallets. 

Todo:

- [x] finish benchmarks
- [x] update code documentation/comments
- [x] integration tests
- [x] runtime integration

### CODE REVIEW

File: README.md
Comment (martinfridrich) [inline README.md:1]: Why did you remove readme?

File: pallets/referrals/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+		asset_id: T::AssetId,
+		amount: Balance,
+	) -> Result<Balance, DispatchError> {
+		// Does trader have a linked referral account ?
+		let Some(ref_account) = Self::linked_referral_account(&trader) else {
+			return Ok(amount);
+		};
+		// What is the referer level?
+		let Some((level,_)) = Self::referrer_level(&ref_account) else {
+			// Should not really happen, the ref entry should be always there.
```
Comment (martinfridrich) [inline pallets/referrals/src/lib.rs:518]: I would add defensive error so it's logged and we know something went wrong

File: pallets/referrals/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+		let Some(price) = T::SpotPriceProvider::spot_price(T::RewardAsset::get(), asset_id) else {
+			// no price, no fun.
+			return Ok(amount);
+		};
+
+		let referrer_reward = tier.referrer.mul_floor(amount);
+		let trader_reward = tier.trader.mul_floor(amount);
+		let total_taken = referrer_reward.saturating_add(trader_reward);
+		ensure!(total_taken <= amount, Error::<T>::IncorrectRewardCalculation);
+		T::Currency::transfer(asset_id, &source, &Self::pot_account_id(), total_taken, true)?;
```
Comment (martinfridrich) [inline pallets/referrals/src/lib.rs:536]: What happen if `total_taken` is less than ED and `pot` is empty? non-dustable whitelist doesn't work for pallet-balances and HDX.

File: pallets/referrals/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+		/// Register new referral code.
+		///
+		/// `origin` pays the registration fee.
+		/// `code` is assigned to the given `account`.
+		///
+		/// Length of the `code` must be at least `MIN_CODE_LENGTH`.
+		/// Maximum length is limited to `T::CodeLength`.
+		/// `code` must contain only alfa-numeric characters and all characters will be converted to upper case.
+		///
+		/// /// Parameters:
```
Comment (martinfridrich) [inline pallets/referrals/src/lib.rs:279]:
```suggestion
		/// Parameters:
```

File: pallets/referrals/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+//! used their referral code to link their account to the referrer account.
+//!
+//! The trader can get back part of the trade fee too if configured.
+//!
+//! Pallet also provides support for volume-based tiering. Referrer can reached higher Level based on the total amount generated by users of the referrer code.
+//! The higher level, the better reward.
+//!
+//! Rewards are accumulated in the pallet's account and if it is not RewardAsset, it is converted to RewardAsset prior to claim.
+//!
+//! //! ### Terminology
```
Comment (martinfridrich) [inline pallets/referrals/src/lib.rs:32]:
```suggestion
//! ### Terminology
```

File: pallets/referrals/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+				let reward_reserve_hp = U256::from(reward_reserve);
+				let share_issuance_hp = U256::from(share_issuance);
+				let r = shares_hp
+					.checked_mul(reward_reserve_hp)?
+					.checked_div(share_issuance_hp)?;
+				Balance::try_from(r).ok()
+			}()
+			.ok_or(ArithmeticError::Overflow)?;
+
+			T::Currency::transfer(T::RewardAsset::get(), &Self::pot_account_id(), &who, rewards, true)?;
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:421]: i will change this. But wuold it fail even if the account is excluded from dusting ? just curious.

File: pallets/referrals/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+			let w  = <T as Config>::WeightInfo::claim_rewards();
+			let one_read = T::DbWeight::get().reads(1_u64);
+			w.saturating_add(convert_weight).saturating_add(one_read)
+		})]
+		pub fn claim_rewards(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			for (asset_id, _) in Assets::<T>::drain() {
+				let asset_balance = T::Currency::balance(asset_id, &Self::pot_account_id());
+				T::Convert::convert(Self::pot_account_id(), asset_id, T::RewardAsset::get(), asset_balance)?;
+			}
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:401]: I hope the TX would not even execute. 

In this case, convert needs to be called first.

File: pallets/referrals/src/lib.rs (hunk 2 above)
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:536]: Right... good point. let me think about this one

File: pallets/referrals/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinfridrich) [reply pallets/referrals/src/lib.rs:421]: orml-tokens no, balances yes, hdx accounts can't be excluded from dusting

File: pallets/referrals/src/lib.rs (hunk 2 above)
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:549]: we added a seed amount.

File: integration-tests/src/referrals.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+use hydradx_runtime::{Currencies, Omnipool, Referrals, Runtime, RuntimeOrigin, Tokens};
+use orml_traits::MultiCurrency;
+use primitives::AccountId;
+use sp_runtime::Permill;
+use xcm_emulator::TestExt;
+
+#[test]
+fn registering_a_code_should_charge_registration_fee() {
+	Hydra::execute_with(|| {
+		let code = b"BALLS69".to_vec();
```
Comment (jak-pan) [inline integration-tests/src/referrals.rs:14]: LOL

Comment (enthusiastmartin) [reply integration-tests/src/referrals.rs:14]: that's @mrq1911 desired code.

File: pallets/referrals/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+		},
+		/// Referral code has been linked to an account.
+		CodeLinked {
+			account: T::AccountId,
+			code: ReferralCode<T::CodeLength>,
+			referral_account: T::AccountId,
+		},
+		/// Asset has been converted to RewardAsset.
+		Converted {
+			from: T::AssetId,
```
Comment (mrq1911) [inline pallets/referrals/src/lib.rs:237]: probably it would make sense to have "Amount" struct with asset id and value, so you would have only from & to fields here : Amount

File: pallets/referrals/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+		/// Maximum length is limited to `T::CodeLength`.
+		/// `code` must contain only alfa-numeric characters and all characters will be converted to upper case.
+		///
+		/// Parameters:
+		/// - `code`: Code to register. Must follow the restrictions.
+		///
+		/// Emits `CodeRegistered` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::register_code())]
+		pub fn register_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
```
Comment (mrq1911) [inline pallets/referrals/src/lib.rs:297]: cannot you limit length of the string on the input as well ?

File: pallets/referrals/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+		///
+		/// IF there is any asset in the reward pot, all is converted to RewardCurrency first.
+		///
+		/// Reward amount is calculated based on the shares of the signer account.
+		///
+		/// if the signer account is referrer account, total accumulated rewards is updated as well as referrer level if reached.
+		///
+		/// Emits `Claimed` event when successful.
+		#[pallet::call_index(3)]
+		#[pallet::weight( {
```
Comment (mrq1911) [inline pallets/referrals/src/lib.rs:406]: wow, nice, i didnt know this is possible... check it out @dmoka

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: pallets/referrals/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+	Advanced,
+	Expert,
+}
+
+impl Level {
+	pub fn next_level(&self) -> Self {
+		match self {
+			Self::Novice => Self::Advanced,
+			Self::Advanced => Self::Expert,
+			Self::Expert => Self::Expert,
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:93]: Why wrong ?

File: pallets/referrals/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+pub enum Level {
+	#[default]
+	Novice,
+	Advanced,
+	Expert,
+}
+
+impl Level {
+	pub fn next_level(&self) -> Self {
+		match self {
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:90]: you could.

File: runtime/hydradx/src/assets.rs [hunk 13]
```diff
@@ -985,3 +992,141 @@ impl pallet_xyk::Config for Runtime {
 	type NonDustableWhitelistHandler = Duster;
 	type OracleSource = XYKOracleSourceIdentifier;
 }
+
+parameter_types! {
+	pub const ReferralsPalletId: PalletId = PalletId(*b"referral");
+	pub RegistrationFee: (AssetId,Balance, AccountId)= (NativeAssetId::get(), 222_000_000_000_000, TreasuryAccount::get());
+	pub const MaxCodeLength: u32 = 7;
+	pub const ReferralsOraclePeriod: OraclePeriod = OraclePeriod::Short;
```
Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:1001]: i think that short is too short, ten minutes or even longer seems to be safer - the oracle period should roughly match how often we will be converting it, and that will be more than every 20 blocks

File: runtime/hydradx/src/assets.rs [hunk 14]
```diff
@@ -985,3 +992,141 @@ impl pallet_xyk::Config for Runtime {
...
+	type AssetId = AssetId;
+	type Currency = FungibleCurrencies<Runtime>;
+	type Convert = ConvertViaOmnipool;
+	#[cfg(not(feature = "runtime-benchmarks"))]
+	type PriceProvider =
+		OraclePriceProviderUsingRoute<Router, OraclePriceProvider<AssetId, EmaOracle, LRNA>, ReferralsOraclePeriod>;
+	#[cfg(feature = "runtime-benchmarks")]
+	type PriceProvider = ReferralsDummyPriceProvider;
+	type RewardAsset = NativeAssetId;
+	type PalletId = ReferralsPalletId;
```
Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:1017]: any reason why we dont use default pallet id?

Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:1017]: i would probably set some min based on oracle atleast, so you cannot rug it

File: runtime/hydradx/src/assets.rs [hunk 15]
```diff
@@ -985,3 +992,141 @@ impl pallet_xyk::Config for Runtime {
...
+	type Error = DispatchError;
+
+	fn convert(
+		who: AccountId,
+		asset_from: AssetId,
+		asset_to: AssetId,
+		amount: Balance,
+	) -> Result<Balance, Self::Error> {
+		let balance = Currencies::free_balance(asset_to, &who);
+		Omnipool::sell(RuntimeOrigin::signed(who.clone()), asset_from, asset_to, amount, 0)?;
```
Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:1038]: i would probably set some limit based on oracle just to be safe

File: pallets/referrals/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply pallets/referrals/src/lib.rs:93]: next level is the same level

[the remark this answers is not in the document]
Comment (mrq1911) [reply pallets/referrals/src/lib.rs:93]: how do you represent you reached max level?

[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:93]: for the max level, is the max level - this is only used when we try to figure out if it should be moved. and we check if the top level is already reached.

File: pallets/referrals/src/lib.rs (hunk 9 above)
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:297]: I could used bounded vec for param too. But in the polkadot.js when i tested it - it throws some weird errors when i try to enter longer code. something about wasm unreachable.

File: runtime/hydradx/src/assets.rs (hunk 15 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1038]: How could I determine such limit ?

File: runtime/hydradx/src/assets.rs (hunk 14 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1017]: we always did this for other pallets, just did the same.

File: runtime/hydradx/src/assets.rs (hunk 13 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs]: changed to 10mints. We dont have longer one in hydra

File: pallets/referrals/src/lib.rs (hunk 9 above)
Comment (mrq1911) [reply pallets/referrals/src/lib.rs:297]: thats good tho, you shouldn't input longer code :)

File: runtime/hydradx/src/assets.rs (hunk 15 above)
Comment (mrq1911) [reply runtime/hydradx/src/assets.rs:1038]: something like short price - 1% ?

File: pallets/referrals/src/lib.rs (hunk 9 above)
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:297]: changed to bounded vec

File: runtime/hydradx/src/assets.rs (hunk 15 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1038]: added

File: pallets/referrals/src/lib.rs (hunk 8 above)
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs]: done

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: runtime/hydradx/src/assets.rs [hunk 16]
```diff
@@ -985,3 +994,207 @@ impl pallet_xyk::Config for Runtime {
...
+			Level::Tier4 => Tier {
+				referrer: Permill::from_percent(25),
+				trader: Permill::from_percent(15),
+				external: Permill::from_percent(10),
+			},
+		}
+	}
+}
+
+#[cfg(feature = "runtime-benchmarks")]
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1124]: we have other stuff the same way here in this file.

[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/assets.rs:1124]: should probably be closer to the benchmarks than the production runtime

File: pallets/referrals/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,674 @@
...
+			if amount >= required {
+				return next_level.increase::<T>(amount);
+			}
+			self
+		}
+	}
+}
+
+#[derive(Clone, Copy, Default, Encode, Decode, Eq, PartialEq, RuntimeDebug, MaxEncodedLen, TypeInfo)]
+pub struct Tier {
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:124]: it is not that easy. we need a way to determine referrer level and based on that we obtain reward percentages. with one, it is not really possible.

File: pallets/referrals/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,674 @@
...
+		pub fn claim_rewards(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			for (asset_id, _) in Assets::<T>::drain() {
+				let asset_balance = T::Currency::balance(asset_id, &Self::pot_account_id());
+				let r = T::Convert::convert(Self::pot_account_id(), asset_id, T::RewardAsset::get(), asset_balance);
+				if let Err(error) = r {
+					if error == Error::<T>::ConversionMinTradingAmountNotReached.into()
+						|| error == Error::<T>::ConversionZeroAmountReceived.into()
+					{
+						// We allow these errors to continue claiming as the current amount of asset that needed to be converted
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:464]: actually, there is a little mistake in here,as i cannot drain all assets anymore. in case of error that we ignore, we need to keep it in the storage.  I need to rework this sligtly.

File: runtime/hydradx/src/assets.rs [hunk 19]
```diff
@@ -985,3 +994,207 @@ impl pallet_xyk::Config for Runtime {
...
+			Level::Tier1 => 1_222_222_000_000_000_000,
+			Level::Tier2 => 12_222_220_000_000_000_000,
+			Level::Tier3 => 122_222_200_000_000_000_000,
+			Level::Tier4 => 1_222_222_000_000_000_000_000,
+		}
+	}
+}
+
+pub struct ReferralsTierRewards;
+impl GetByKey<Level, Tier> for ReferralsTierRewards {
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1087]: we could have just one config param which returns both - percentages and volume per given level. But as iti used - you need one or another, not both at the same time. 

well, it does not really matter - we can just ignore what we dont need.

File: runtime/hydradx/src/assets.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1124]: i would refactor all this as separeate task.

File: pallets/referrals/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,674 @@
...
+	#[pallet::storage]
+	#[pallet::getter(fn asset_tier)]
+	pub(super) type AssetTier<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::AssetId, Blake2_128Concat, Level, Tier, OptionQuery>;
+
+	/// Information about assets that are currently in the rewards pot.
+	/// Used to easily determine list of assets that need to be converted.
+	#[pallet::storage]
+	#[pallet::getter(fn assets)]
+	pub(super) type Assets<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, ()>;
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs]: renamed to PendingConversions

File: runtime/hydradx/src/assets.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs]: done

File: pallets/referrals/src/lib.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:124]: renamed Tier and kept the Level and FeeDistribution types as is.

File: pallets/referrals/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,674 @@
...
+				rewards,
+				keep_pot_alive,
+			)?;
+			TotalShares::<T>::mutate(|v| {
+				*v = v.saturating_sub(shares);
+			});
+			Referrer::<T>::mutate(who.clone(), |v| {
+				if let Some((level, total)) = v {
+					*total = total.saturating_add(rewards);
+					*level = level.increase::<T>(*total);
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:506]: added

File: runtime/hydradx/src/assets.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/assets.rs:1118]: https://github.com/galacticcouncil/HydraDX-node/issues/722

File: pallets/referrals/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply pallets/referrals/src/lib.rs:463]: i though that next trade of tat asset will just add it back to the list, so you have chance it will have sufficient amount that time

no need to retry a conversion that failed before right?

File: runtime/hydradx/src/assets.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (mrq1911) [reply runtime/hydradx/src/assets.rs]: There are still no HDX rewards (all those should give 0 to the referrals - everything goes to the staking instead)

File: runtime/hydradx/src/assets.rs [hunk 22]
```diff
@@ -1004,3 +1013,201 @@ impl pallet_xyk::Config for Runtime {
...
+	}
+}
+
+pub struct ReferralsLevelVolumeAndRewards;
+
+impl GetByKey<Level, (Balance, FeeDistribution)> for ReferralsLevelVolumeAndRewards {
+	fn get(k: &Level) -> (Balance, FeeDistribution) {
+		let volume = match k {
+			Level::Tier0 | Level::None => 0,
+			Level::Tier1 => 1_222_222_000_000_000_000,
```
Comment (mrq1911) [inline runtime/hydradx/src/assets.rs:1096]: please make sure these numbers are final

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: pallets/referrals/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply pallets/referrals/src/lib.rs:463]: it will retry when the asset is back inthe list. now it is all good

File: runtime/hydradx/src/assets.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs]: done. if HDX, it does not go through referrals.

File: runtime/hydradx/src/assets.rs (hunk 22 above)
Comment (enthusiastmartin) [reply runtime/hydradx/src/assets.rs:1096]: adjusted

**Review by mrq1911 — APPROVED** [review]
